chaste_codegen/rl_model.py

```diff
@@ -0,0 +1,139 @@
+from chaste_codegen._partial_eval import partial_eval
+from chaste_codegen.chaste_model import ChasteModel
+
+
+class RlModel(ChasteModel):
+    """ Holds template and information specific for the RushLarsen model type"""
+
+    def __init__(self, model, file_name, **kwargs):
+        super().__init__(model, file_name, **kwargs)
+        self._derivative_equations = \
+            partial_eval(self._derivative_equations, self._y_derivatives, keep_multiple_usages=False)
+        self._non_linear_state_vars = \
+            get_non_linear_state_vars(self._derivative_equations, self._membrane_voltage_var,
+                                      self._state_vars, self._printer)
+
+        self._derivative_alpha_beta, self._formatted_alpha_beta_eqs, self._vars_in_derivative_alpha_beta = \
+            self._get_formatted_alpha_beta()
+        self._formatted_alpha_beta_eqs = self._format_alpha_beta_eqs()
+
+    def _get_formatted_alpha_beta(self):
+        """Gets the information for rAlphaOrTau, rBetaOrInf and formatted equations"""
+        def match_alpha_beta(expr, x):  # expr already in piecewise_fold form
+            """Match alpha*(1-x) - beta*x"""
+            a, b = None, None
+            alpha = sp.Wild('alpha', exclude=[x])
+            beta = sp.Wild('beta', exclude=[x])
+            match = expr.expand().match((alpha - x * alpha) - beta * x)
+            if match is not None:
+                a, b = match[alpha], match[beta]
+            return {'alpha': a, 'beta': b}
+
+        def match_inf_tau(expr, x):  # expr already in piecewise_fold form
+            """Match (inf-x)/tau"""
+            i, t = None, None
+            inf = sp.Wild('inf', exclude=[x])
+            tau = sp.Wild('tau', exclude=[x])
+            match = expr.expand().match(inf / tau - x / tau)
+            if match is not None:
+                i, t = match[inf], match[tau]
+            return {'inf': i, 'tau': t}
+
+        derivative_alpha_beta, vars_in_derivative_alpha_beta = [], set()
+
+        # Substitute non-linear bits into derivative equations, so that we can pattern match
+        linear_derivs_eqs = derives_eqs_partial_eval_non_linear(self._y_derivatives, self._non_linear_state_vars,
+                                                                self._membrane_voltage_var, self._state_vars,
+                                                                self._get_equations_for)
+
+        for deriv in self._y_derivatives:
+            ab = {'alpha': None}
+            it = {'tau': None}
+            # get match if possible (deiv is linear)
+            if deriv.args[0] not in self._non_linear_state_vars and deriv.args[0] is not self._membrane_voltage_var:
+                eq = [e for e in linear_derivs_eqs if e.lhs == deriv][0]
+                ab = match_alpha_beta(eq.rhs, eq.lhs.args[0])
+                it = match_inf_tau(eq.rhs, eq.lhs.args[0])
```

Comment:
You could do this only if ab isn't a match?

Reply:
I don't really understand what you mean? This bit is to figure out whether there is a match first. I do need the elaborate if as otherwise the equation doesn't exist. I could possible do the eq check outside minus the index and just check len>0 in the if.

Reply:
As in, only check for an inf_tau match in this case. But sounds like you figured that out already :)